Performance improvements

akoebbe opened the pull request:
I was getting PHP timeouts when I was hitting the admin dashboard. So I've come up with some performance enhancements.

Changes proposed in this pull request:

- Improved memory consumption and speed for the dashboard
  - Changed from pulling the entire log in to memory, to using the `SplFileObject` and looping through it.
  - Pulled a $hostname check out of iteration since it only needs to be looked up once.
  - Found a more efficient way of calculating the number of domains in gravity.list.
- Removed the callback functions and swapped them out for `foreach` so we can iterate on the `SplFileObject`.
- Fixed a bug where you would get a error on `max()` when you had no Ad records in your log file.

Before the change:
<img width="1347" alt="pi-hole_devel" src="https://cloud.githubusercontent.com/assets/1044465/19217643/a7e7d142-8da5-11e6-8e80-bd6c6780e3de.png">

After the change:
<img width="1110" alt="pi-hole_improvments" src="https://cloud.githubusercontent.com/assets/1044465/19217644/adecdb0a-8da5-11e6-8f81-b7b2faf8d276.png">


@pi-hole/dashboard

Mcat12 commented:
That's a big improvement! Currently we're also implementing a database approach to the performance issue, but I think most of your changes will still be useful.

Mcat12 commented:
Why is that merge commit in this PR?

akoebbe commented:
Good question. Probably because I originally branched from master and I
tried to correct that mistake. I'll get that corrected. I also found a bug
with the "over time" API that I introduced. I'll fix that up as well.

On Oct 9, 2016 7:32 AM, "Mcat12" <notifications@github.com> wrote:

> Why is that merge commit in this PR?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pi-hole/AdminLTE/pull/156#issuecomment-252483904>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA_v8TqBAfqNYtO45SZb_zOo4nLk-JZLks5qyN7ugaJpZM4KR5Gv>
> .
>

akoebbe commented:
Ok I've fixed up my branch to only include my changes and squashed the commits while I was at it. I totally agree with the DB solution as the end goal. This just seems to help the current situation until the other is ready. I would like to have a few people try it out in their pihole environments to make sure this really helps the audience at large.

Mcat12 commented:
Approved

PromoFaux commented:
Approved

I like this as an intermediate step before `dbintegrate`, and the gravity count will come in handy there, too.
